qiskit/circuit/controlflow/for_loop.py

```diff
@@ -116,3 +116,101 @@ def params(self, parameters):
     @property
     def blocks(self):
         return (self._params[2],)
+
+
+class ForLoopContext:
+    """A context manager for building up ``for`` loops onto circuits in a natural order, without
+    having to construct the loop body first.
+
+    Within the block, a lot of the bookkeeping is done for you; you do not need to keep track of
+    which qubits and clbits you are using, for example, and a loop parameter will be allocated for
+    you, if you do not supply one yourself.  All normal methods of accessing the qubits on the
+    underlying :obj:`~QuantumCircuit` will work correctly, and resolve into correct accesses within
+    the interior block.
+
+    You generally should never need to instantiate this object directly.  Instead, use
+    :obj:`.QuantumCircuit.for_loop` in its context-manager form, i.e. by not supplying a ``body`` or
+    sets of qubits and clbits.
+
+    Example usage::
+
+        import math
+        from qiskit import QuantumCircuit
+        qc = QuantumCircuit(2, 1)
+
+        with qc.for_loop(None, range(5)) as i:
+            qc.rx(i * math.pi/4, 0)
+            qc.cx(0, 1)
+            qc.measure(0, 0)
+            qc.break_loop().c_if(0)
+
+    This context should almost invariably be created by a :meth:`.QuantumCircuit.for_loop` call, and
+    the resulting instance is a "friend" of the calling circuit.  The context will manipulate the
```

Comment:
(C++ nitpicking) I think for the purpose of user-facing docs, `friend` here is more of an implementation detail, i.e. for the user, it's not important that the underlying API this context uses to modify the originating circuit is private.

Reply:
Yeah, for sure. As it stands, this documentation isn't user-facing, though, only developer-facing - it's not built into the regular API docs. I was trying to document implementation as well in these cases.

There's very little actual public documentation in this PR. I think it's just the `if_test`, `for_loop` and `while_loop` methods of `QuantumCircuit` that actually face the public. The rest of the documentation for these will go in tutorials, I think.